File: java/src/main/java/shapes/base/models/Circle.java
package shapes.base.models;

import java.util.List;
import java.util.Map;
import java.util.stream.Collectors;

import shapes.base.commons.BaseShape;
import shapes.base.commons.Dimensional;
import shapes.base.commons.Displayable;

public class Circle extends BaseShape implements Dimensional, Displayable {
    private final double radius;
    private String color;

    public Circle(double radius) {
        double dim1 = (radius > 0) ? radius : 1;

        super(
            List.of(
                Map.of("radius", Math.max(radius, 1)),
                Map.of("diameter", Math.max(radius, 1) * 2),
                Map.of("curvature", 1 / Math.max(radius, 1))
            )
        );

        this.radius = dim1;
        this.color = "None";
    }

    public String getColor() {
        return color;
    }

    public void setColor(String color) {
        this.color = color;
    }

    @Override
    public void printDescription() {
        System.out.println("Circle radius: " + radius);
        System.out.println("Circle dimensions: " + formatDimensions());
        System.out.println("Circle color: " + color);
        System.out.println("Circle area: " + area());
        System.out.println("Circle perimeter: " + perimeter());
        System.out.println("Circle is displayable: " + shouldDisplay());
        System.out.println("Circle is a shape: " + Circle.validateShape());
    }

    @Override
    public double area() {
        return Math.PI * Math.pow(radius, 2);
    }

    @Override
    public double perimeter() {
        return 2 * Math.PI * radius;
    }

    private String formatDimensions() {
        List<String> parts = getDimensions()
            .stream()
            .map(
                dim -> dim
                    .entrySet()
                    .stream()
                    .map(e -> e.getKey() + ": " + String.format("%.3f", e.getValue()))
                    .collect(Collectors.joining(", "))
            )
            .collect(Collectors.toList());

        return String.join(" | ", parts);
    }
}

File: scala/src/main/scala/shapes/base/models/Circle.scala
package shapes.base.models

import shapes.base.commons.{BaseShape, Dimensional, Displayable}

import scala.compiletime.uninitialized
import scala.math.Pi

final class Circle(radius: Double)
    extends BaseShape(
      List(
        Map("radius" -> Math.max(radius, 1.0)),
        Map("diameter" -> (Math.max(radius, 1.0) * 2)),
        Map("curvature" -> (1.0 / Math.max(radius, 1.0)))
      )
    )
    with Dimensional
    with Displayable {
    private var _color: String = "None"

    def color: String = _color
    def color_=(value: String): Unit = _color = value

    override def printDescription(): Unit = {
        println(s"Circle radius: $radius")
        println(s"Circle dimensions: ${this.formatDimensions}")
        println(s"Circle color: ${this.color}")
        println(s"Circle area: ${this.area}")
        println(s"Circle perimeter: ${this.perimeter}")
        println(s"Circle is displayable: ${this.shouldDisplay}")
        println(s"Circle is a shape: ${BaseShape.validateShape}")
    }

    override def area: Double = {
        Pi * Math.pow(radius, 2)
    }

    override def perimeter: Double = {
        2 * Pi * radius
    }

    private def formatDimensions: String = {
        val parts: List[String] = dimensions.map(
          _.map { case (key, value) => s"${key.toUpperCase}: ${"%.3f".format(value)}" }
              .mkString(" | ")
        )

        parts.mkString(" | ")
    }
}
